lib/sales_engine.rb

```diff
@@ -153,145 +106,128 @@ def find_invoice_from_transaction(id)
     end
-    merchant_items
   end
 
-  def find_most_revenue_from_merchant_repository(x)
-    merchant_ids = merchant_revenue.sort[-x..-1].collect {|i| i[1]}
-    merchants = merchant_ids.map do |merchant_id|
-      merchant_repository.find_by_merchant_id(merchant_id)
-    end
-    merchants.reverse
+  def find_most_revenue_from_merchant_repo(x)
+    merchant_ids = merchant_revenue.sort[-x..-1].map { |i| i[1] }.reverse
+    find_merchants_from(merchant_ids)
   end
 
-  def find_most_items_sold_from_merchant_repository(x)
-    merchant_ids = merchant_items.sort[-x..-1].collect {|i| i[1]}
-    merchants = merchant_ids.map do |merchant_id|
-      merchant_repository.find_by_merchant_id(merchant_id)
-    end
-    merchants.reverse
+  def find_most_items_sold_from_merchant_repo(x)
+    merchant_ids = merchant_items.sort[-x..-1].map { |i| i[1] }.reverse
+    find_merchants_from(merchant_ids)
   end
 
-  def find_revenue_by_date_from_merchant_repository(date)
-    merchant_revenue(date).collect {|i| i[0]}.reduce(:+)
+  def find_revenue_by_date_from_merchant_repo(date)
+    merchant_revenue(date).map { |i| i[0] }.reduce(:+)
   end
 
-  def find_most_revenue_items(x)
-
-    all_invoices = invoice_repository.all
-      invoice_ids = all_invoices.map(&:id)
-    all_transactions = invoice_ids.map do |invoice_id|
-      transaction_repository.find_all_by_invoice_id(invoice_id)
-    end
-
-    successful_ids = all_transactions.flatten.map do |transaction|
-      transaction.invoice_id unless transaction.result != ("success")
-    end
-
-    invoice_items = successful_ids.map do |invoice_id|
-      invoice_item_repository.find_all_by_invoice_id(invoice_id)
-    end
-
-    items_revenue = invoice_items.flatten.map do |id|
-      [id.quantity * id.unit_price,id.item_id ]
+  def find_successful_invoice_ids_from_transactions(transactions=transaction_repository.all)
+    transactions.flatten.map do |transaction|
+      transaction.invoice_id if transaction.result == ("success")
     end
+  end
 
-    item_hash = items_revenue.group_by{ |n| n[1] }
-    item_revenue = []
-    for n in 1..10000
-
-      unless item_hash[n].nil?
-        item_revenue << [item_hash[n].collect{ |n| n[0] }.reduce(:+), item_hash[n].flatten[1]]
-      end
+  def items_with_revenue
+    invoice_ids = find_successful_invoice_ids_from_transactions
+    invoice_items = find_invoice_items_from_invoice_ids(invoice_ids)
+    items_revenue = Hash.new(0)
+    invoice_items.flatten.map do |id|
+      items_revenue[id.item_id] += id.quantity * id.unit_price
     end
+    items_revenue
+  end
 
-    item_ids = item_revenue.sort[-x..-1].collect {|i| i[1]}
-    items = item_ids.map do |item_id|
+  def find_most_revenue_items(x)
+    all_item_ids = items_with_revenue.sort_by { |_, val| -val }.map { |i| i[0] }
+    all_item_ids[0...x].map do |item_id|
       item_repository.find_by_item_id(item_id)
     end
-    return items.reverse
+  end
+
+  def find_most_popular_items(x)
```

Comment:
Unused method argument - `x`. If it's necessary, use `_` or `_x` as an argument name to indicate that it won't be used. You can also write as `find_most_popular_items(*)` if you want the method to accept any arguments but don't care about them.